R/sel.R

```diff
@@ -97,3 +96,100 @@ m_sel <- function(
 
   m_sel
 }
+
+#' Selection Across Multiple Residues
+#'
+#' Behaves just like the \code{m_sel()}, but returns a new selection for each
+#' residue specified with \code{resi}.
+#'
+#' The \code{m_sel(resi = 1:10)} returns a selection of all 10 residues.
+#' The \code{m_multiresi_sel(resi = 1:10)} returns 10 individual selections,
+#' each containing only 1 of the residues.
+#'
+#' @param model a single model or list of models from which atoms should be
+#' selected. Can also specify by numerical creation order.
+#' Reverse indexing is allowed (-1 specifies last added model).
+#' @param resi Residue number/s. (vector)
+#' @param resn Parent residue name as 3-letter code (e.g. "ALA", "GLY",
+#' "CYS"...)
+#' @param bonds overloaded to select number of bonds, e.g. \code{bonds = 0}
+#' will select all non-bonded atoms
+#' @param invert Logical, if \code{invert = TRUE}, Inverts the selection
+#' criteria.
+#' @param byres Logical, if \code{byres = TRUE}, expands the selection to
+#' entire residues that include any selected atoms.
+#' @param expand Expand selection to include atoms within a specified distance
+#' from current selection.
+#' all atoms of any residue that has any atom already selected.
+#' @param elem element abbreviation (e.g 'H', 'Ca', etc)
+#' @param chain String, chain this atom belongs to (e.g. 'A' for chain A)
+#' @param atom Atom name, may be more specific than 'elem' (e.g. 'CA' for alpha
+#' carbon)
+#' @param ss Secondary structure identifier. 'h' for helix, 's' for beta-sheet.
+#' @param b Atom b factor data
+#' @param clickable Set this flag to true to enable click selection handling
+#' for this atom
+#' @param callback Callback click handler function to be executed on this atom
+#' and its parent viewer.
+#'
+#' @return sel \code{list()} for selecting atoms.
+#' @export
+#'
+#' @examples
+#' library(r3dmol)
+#'
+#' r3dmol() %>%
+#'   m_add_model(data = pdb_6zsl) %>%
+#'   m_set_style(style = m_style_cartoon()) %>%
+#'   m_zoom_to() %>%
+#'   m_add_style(
+#'     sel = m_sel(resi = 1:10),
+#'     style = c(
+#'       m_style_stick(),
+#'       m_style_sphere(scale = 0.3)
+#'     )
+#'   ) %>%
+#'   m_add_lines(
+#'     starts = m_sel(resi = 1, chain = "A"),
+#'     ends = m_multiresi_sel(
+#'       resi = 2:10,
+#'       chain = "B"
```

Comment:
What would be happened if using the method as below?
```r
m_multiresi_sel(
  resi = 2:10,
  chain = c("A", "B")
)
```

Reply:
Currently I have it setup that it would would return:
```r
list(
  m_sel(resi = 2, 
        chain = c("A", "B")), 
  m_sel(resi = 3, 
        chain = c("A", "B")), 
  etc...
)
```

I considered having it being able to return combinations of the other options as well, but figured that it would be too complex and not that useful. 

When doing something like drawing lines etc, usually it is between residues, so you might want to select lots of different residues. Less so that you want to select lots of combinations of other things. 

Rather than creating a function for every possible variation, I figured it would be best just to let the user call multiple functions, as it will be less common than wanting to selecting multiple residues.

Reply:
I sorry about that I should write the comment more clear. Yes, we don't need to return all the combinations of all options. I mean that all options are pass to the JavaScript side directly without any handling. If the 3Dmol.js is only accept a string as parameter and we pass a list here, will it render correctly (according to the [docs](http://3dmol.csb.pitt.edu/doc/types.html#AtomSpec), the chain only accept a string parameter)? 

I have no idea how to test the problem, the only way that I have known is create another pkgdown article, tell the user how to use the function, we could check the render results, too.

Reply:
I'm unsure what you mean. In all of the testing that I am doing it seems to be working OK. 

For example, the following returns lists of selections and they are passed on OK:
```r
r3dmol() %>% 
  m_add_model(m_fetch_pdb("1bna")) %>% 
  m_set_style(m_style_cartoon()) %>% 
  m_add_lines(
    starts = m_multiresi_sel(resi = 1:3, chain = "A"), 
    ends = m_multiresi_sel(resi = 22:20, chain = "B")
  ) %>% 
  m_zoom_to()
```
<p align="center">
<img src="https://user-images.githubusercontent.com/36021261/108949566-bd7bcf80-76b8-11eb-97f3-f2af1599e6b6.png" width="400">
</p>

You can see that there are 3 selections for each chain, and a line between them each.

Is this what you are concerned about? I have ensured that it handles the correct level of the list properly before passing it off to the javascript call.

Reply:
I am very sorry that my English is not good enough to explain clearly what I want to explain.
For example, if you use the following code, you will get the same result as in your screenshot:
```r
r3dmol() %>% 
    m_add_model(m_fetch_pdb("1bna")) %>% 
    m_set_style(m_style_cartoon()) %>% 
    m_add_lines(
        starts = m_multiresi_sel(resi = 1:3, chain = c("A", "B")), 
        ends = m_multiresi_sel(resi = 22:20, chain = "B")
    ) %>% 
    m_zoom_to()
```
If the user want to set 6 lines as expected but only show 3 here, there is also no relevant alert message, which may let the user considered as a bug.

Reply:
Okay I see what you mean now. Don't apologise, I can only speak one language so you're doing a lot better than me! :)

I realised that the `1bna` example isn't a great example. `Chain A` contains residues 1-18 and `chain B` contains residues 19 - 36.

In this example below, slightly modified code but the idea is the same:
```r
r3dmol() %>% 
  m_add_model(pdb_6zsl) %>% 
  m_set_style(m_style_cartoon(color = "lightgreen"), 
              m_sel(chain = "A")) %>% 
  m_set_style(m_style_cartoon(color = "red"), 
              m_sel(chain = "B")) %>% 
  m_add_style(m_style_stick(), 
              sel = m_sel(resi = c(2:4, 22:20), chain = c("A", "B"))) %>% 
  m_add_lines(
    starts = m_multiresi_sel(resi = 2:4, chain = c("A", "B")), 
    ends = m_multiresi_sel(resi = 22:20, chain = "B"), 
    color = c("green", "blue", "red")
  ) %>% 
  m_zoom_to()
```
<p align="center">
<img src="https://user-images.githubusercontent.com/36021261/108953339-25cdaf80-76bf-11eb-8c13-64436d2c8978.png" width="400">
</p>

All of the `starts` contain both `chain A` and `chain B` (so originate from the space between the two), whereas the `ends` are all on `chain B` and so go to their correct residues. 

You are correct that a user might expect 6 total lines, but when thinking about how to implement `multi_sel()` style functions, I figured it might be best to keep it simple and just make it clear to the user that it is only creating new selections based on the `resi` argument. That is why I named it `m_multiresi_sel()`. 

Potentially we have multiple `multi_sel` style functions, or if you have a good idea of how to implement the function better. 

I do it this way because it is more common that I will want to select multiple residues, than I will want to select multiple chains. So I am building it more for how I would use it.

Reply:
Thanks for your explanation, I understand what you mean and I can accept this function, but there are two points to consider:
1. function name `m_multiresi_sel()` or `m_multi_resi_sel()`, I perfer the latter but what's your idea?
2. At least we should show some message to tell the user the function only accept multiple resi, and other function only accept string or number. if user pass a vector for other options, we only accept the first one and inform the user about this.

Reply:
Sorry, a little more discussion here:

> All of the starts contain both chain A and chain B (so originate from the space between the two), whereas the ends are all on chain B and so go to their correct residues.

```r
m_add_lines(
    starts = m_multiresi_sel(resi = 2:4, chain = c("A", "B")), 
    ends = m_multiresi_sel(resi = 22:20, chain = "B"), 
    color = c("green", "blue", "red")
  ) %>% 
```

I only seen that 3 lines in the screenshot, so resi 2:4 in chain B is ignored here? (`m_multiresi_sel(resi = 2:4, chain = c("A", "B"))`)

Reply:
Sorry for the confusion!

In the example that you added, the selection below:
```r
m_multiresi_sel(resi = 2:4, chain = c("A", "B"))

# is the same as:
list(
  m_sel(resi = 2, chain = c("A", "B")),
  m_sel(resi = 3, chain = c("A", "B")),
  m_sel(resi = 4, chain = c("A", "B"))
)
```
So all of the start selections contain a single residue (either 2, 3 or 4) but from both chains (chain A (green) & chain B (red)) which is why they start in between the two chains. (see label below, with selection `chain = c("A", "B")`. 

```r
m_add_label(
    sel = m_sel(resi = 2, chain = c("A", "B")), 
    text = "starts location (middle of chains A & B)"
  )
```
<p align="center">

<img src="https://user-images.githubusercontent.com/36021261/108958705-75b07480-76c7-11eb-9c99-0cdec13ec241.png">
</p>

So the selection is including both chains, but because the middle of the selection is in the middle of nowhere, the lines all start from a similar point, which is between their matching residues.

Reply:
This might be more clear: 
```r 
  m_add_label(
    sel = m_sel(resi = 2, chain = c("A", "B")), 
    text = "starts location (middle of chains A & B)"
  ) %>% 
  m_add_label(
    sel = m_sel(resi = 2, chain = c("A")), 
    text = "Chain A, Resi = 2"
  ) %>% 
  m_add_label(
    sel = m_sel(resi = 2, chain = "B"), 
    text = "Chain B, resi = 2")
```

<p align="center">
<img src="https://i.imgur.com/Ku3dUzN.png" width=500>
</p>

Reply:
> 1. function name `m_multiresi_sel()` or `m_multi_resi_sel()`, I perfer the latter but what's your idea?

I agree with this renaming of the function, it makes it clearer as to what it's doing. I'll change it.

> 2. At least we should show some message to tell the user the function only accept multiple resi, and other options only accept string or number. if user pass a vector for other options, we only accept the first one and inform the user about this.

I'll add more documentation, and maybe add a warning if people are passing in other arguments as well.

Reply:
https://github.com/swsoyee/r3dmol/pull/17#discussion_r581670367 https://github.com/swsoyee/r3dmol/pull/17#discussion_r581672964 Oh! I see, understand completely. The function of using the center of all selected points as a new point is convenient, but it is still quite easy to cause misunderstanding, but I think this is the feature of 3dmol itself.

https://github.com/swsoyee/r3dmol/pull/17#discussion_r581673830 Thank you! That would be great. Problem of `m_multi_resi_sel()` has been solved here.

Reply:
> [#17 (comment)](https://github.com/swsoyee/r3dmol/pull/17#discussion_r581670367) [#17 (comment)](https://github.com/swsoyee/r3dmol/pull/17#discussion_r581672964) Oh! I see, understand completely. The function of using the center of all selected points as a new point is convenient, but it is still quite easy to cause misunderstanding, but I think this is the feature of 3dmol itself.

Yep it's the behaviour of 3Dmol itself. Convenient, but will just have to be explicit in documentation to try and convey that to people.